packages/mdc-snackbar/foundation.js

```diff
@@ -0,0 +1,264 @@
+     * @type {number}
+     * @private
+     */
+    this.timeoutMs_ = numbers.DEFAULT_AUTO_DISMISS_TIMEOUT_MS;
+
+    /**
+     * @type {boolean}
+     * @private
+     */
+    this.closeOnEscape_ = true;
+
+    /**
+     * @type {?number}
+     * @private
+     */
+    this.autoDismissTimer_ = null;
+
+    /**
+     * @type {?function(evt: !Event): undefined}
+     * @private
+     */
+    this.transitionEndHandler_ = null;
+  }
+
+  destroy() {
+    const {OPEN, CLOSING} = cssClasses;
+    this.clearAutoDismissTimer_();
+    this.adapter_.removeClass(OPEN);
+    this.adapter_.removeClass(CLOSING);
+  }
+
+  open() {
+    if (this.isOpen) {
+      return;
+    }
+
+    const {OPEN, CLOSING} = cssClasses;
+
+    this.clearAutoDismissTimer_();
+    this.setTransitionEndHandler_(() => {
+      this.adapter_.notifyOpened();
+    });
+    this.autoDismissTimer_ = setTimeout(() => this.close(strings.REASON_DISMISS), this.timeoutMs);
+
+    this.adapter_.addClass(OPEN);
+    this.adapter_.removeClass(CLOSING);
+    this.adapter_.notifyOpening();
+  }
+
+  /**
+   * @param {string=} reason Why the snackbar was closed. Value will be passed to CLOSING_EVENT and CLOSED_EVENT via the
+   *     `event.detail.reason` property. Standard values are REASON_ACTION and REASON_DISMISS, but custom
+   *     client-specific values may also be used if desired.
+   */
+  close(reason = '') {
+    if (!this.isOpen) {
+      return;
+    }
+
+    const {OPEN, CLOSING} = cssClasses;
+
+    this.clearAutoDismissTimer_();
+    this.setTransitionEndHandler_(() => {
+      this.adapter_.removeClass(CLOSING);
+      this.adapter_.notifyClosed(reason);
+    });
+
+    this.adapter_.addClass(CLOSING);
+    this.adapter_.removeClass(OPEN);
+    this.adapter_.notifyClosing(reason);
+  }
+
+  /**
+   * @return {number}
+   */
+  get timeoutMs() {
+    return this.timeoutMs_;
+  }
+
+  /**
+   * @param {number} timeoutMs
+   */
+  set timeoutMs(timeoutMs) {
+    const {
+      MIN_AUTO_DISMISS_TIMEOUT_MS: minValue,
+      MAX_AUTO_DISMISS_TIMEOUT_MS: maxValue,
+    } = numbers;
+
+    if (timeoutMs <= maxValue && timeoutMs >= minValue) {
+      this.timeoutMs_ = timeoutMs;
+    } else {
+      throw new Error(`timeoutMs must be an integer in the range ${minValue}–${maxValue}, but got '${timeoutMs}'`);
+    }
+  }
+
+  /**
+   * @return {boolean}
+   */
+  get closeOnEscape() {
+    return this.closeOnEscape_;
+  }
+
+  /**
+   * @param {boolean} closeOnEscape
+   */
+  set closeOnEscape(closeOnEscape) {
+    this.closeOnEscape_ = closeOnEscape;
+  }
+
+  /**
+   * @return {boolean}
+   */
+  get isOpen() {
+    const {OPEN, CLOSING} = cssClasses;
+    return this.adapter_.hasClass(OPEN) || this.adapter_.hasClass(CLOSING);
+  }
+
+  /**
+   * @param {!Event} evt
+   */
+  handleTransitionEnd(evt) {
+    if (this.transitionEndHandler_) {
+      this.transitionEndHandler_(evt);
+    }
+  }
+
+  /**
+   * @param {!MouseEvent} evt
+   * @private
+   */
+  handleInteraction(evt) {
```

Comment:
As discussed offline, we're going to leave this as-is for now.

We already inspect `event.target` in our wrapper libraries for other components.

Reply:
I thought of one other middle-ground alternative... could we still split this into e.g. `handleActionClick` and `handleIconClick` (going with `click` as per Abhinay's suggestion), and set up the delegated handler in the component? This would leave framework wrappers with the freedom to implement this as they see fit.

Reply:
Done